File: backend/scripts/deploy_contract.py
import os
from pathlib import Path
from dotenv import load_dotenv
import algokit_utils
from algosdk.v2client.algod import AlgodClient
from algosdk.transaction import SuggestedParams
import base64

load_dotenv(Path(__file__).parent.parent.parent / '.env')

def deploy_contracts():
    print("Initializing AlgoKit Client...")
    
    server = os.getenv("ALGOD_SERVER", "https://testnet-api.algonode.cloud")
    port = os.getenv("ALGOD_PORT", "443")
    token = os.getenv("ALGOD_TOKEN", "")
    mnemonic = os.getenv("ALGOD_MNEMONIC")
    
    if not mnemonic:
        raise ValueError("ALGOD_MNEMONIC not set in .env")
        
    algod_client = AlgodClient(token, server)
    account = algokit_utils.get_account_from_mnemonic(mnemonic)
    print(f"Deployer Account: {account.address}")
    
    approval_path = Path(__file__).parent.parent.parent / "contracts" / "build" / "PlacementStudent.approval.teal"
    clear_path = Path(__file__).parent.parent.parent / "contracts" / "build" / "PlacementStudent.clear.teal"
    
    if not approval_path.exists():
        raise FileNotFoundError(f"Cannot find compiled TEAL contracts! Path checked: {approval_path}")
        
    print("Compiling TEAL code...")
    with open(approval_path, "r") as f:
        compiled_approval = algod_client.compile(f.read())
    with open(clear_path, "r") as f:
        compiled_clear = algod_client.compile(f.read())
        
    print("Sending and confirming transaction on TestNet...")
    sp = algod_client.suggested_params()
    
    from algosdk.transaction import ApplicationCreateTxn, StateSchema
    
    txn = ApplicationCreateTxn(
        sender=account.address,
        sp=sp,
        on_complete=0,
        approval_program=base64.b64decode(compiled_approval['result']),
        clear_program=base64.b64decode(compiled_clear['result']),
        global_schema=StateSchema(num_uints=2, num_byte_slices=2),
        local_schema=StateSchema(num_uints=2, num_byte_slices=2)
    )
    
    signed_txn = txn.sign(account.private_key)
    txid = algod_client.send_transaction(signed_txn)
    print(f"Transaction Sent: {txid}")
    
    import algosdk
    confirmation = algosdk.transaction.wait_for_confirmation(algod_client, txid, 4)
    print("\n🚀 CONTRACT SUCCESSFULLY DEPLOYED TO TESTNET 🚀")
    print(f"App ID: {confirmation.get('application-index')}")

if __name__ == "__main__":
    deploy_contracts()

File: backend/scripts/deploy_contract.ts
import * as algokit from '@algorandfoundation/algokit-utils';
import { ApplicationCreateTxn, StateSchema, makeApplicationCreateTxnFromObject, SuggestedParams } from 'algosdk';
import fs from 'fs';
import path from 'path';
import dotenv from 'dotenv';

dotenv.config({ path: '../.env' });

async function deployContracts() {
    // 1. Initialize standard AlgoKit Client for TestNet
    const algodClient = algokit.getAlgoClient({
        server: process.env.ALGOD_SERVER || 'https://testnet-api.algonode.cloud',
        port: process.env.ALGOD_PORT || 443,
        token: process.env.ALGOD_TOKEN || '',
    });

    console.log("AlgoKit Client initialized. Connecting to TestNet...");

    // 2. Fetch the deployment account from environment mnemonics
    if (!process.env.ALGOD_MNEMONIC) {
        throw new Error("Missing ALGOD_MNEMONIC in environment variables!");
    }
    const deployer = await algokit.getAccount(
        { initialFunds: algokit.microAlgos(0) }, // Don't fund on testnet setup automatically
        algodClient
    );
    // Overwrite with actual mnemonic
    const account = algokit.mnemonicAccount(process.env.ALGOD_MNEMONIC);

    console.log(`\nDeployer Account Extracted: ${account.addr}`);

    // 3. Read TEAL contracts compiled by Puya
    const approvalPath = path.resolve(__dirname, '../../contracts/build/PlacementStudent.approval.teal');
    const clearPath = path.resolve(__dirname, '../../contracts/build/PlacementStudent.clear.teal');

    if (!fs.existsSync(approvalPath)) {
        throw new Error(`Cannot find compiled TEAL contracts! Make sure to run 'algokit build'.\nPath checked: ${approvalPath}`);
    }

    const approvalProgram = fs.readFileSync(approvalPath, 'utf8');
    const clearProgram = fs.readFileSync(clearPath, 'utf8');

    console.log("Compiling TEAL code via AlgoKit Utils...");
    const compiledApproval = await algodClient.compile(approvalProgram).do();
    const compiledClear = await algodClient.compile(clearProgram).do();

    // 4. Construct Transaction Parameters using AlgoKit
    const suggestedParams = await algodClient.getTransactionParams().do();
    
    const txnObject = {
        from: account.addr,
        suggestedParams,
        onComplete: 0,
        approvalProgram: new Uint8Array(Buffer.from(compiledApproval.result, 'base64')),
        clearProgram: new Uint8Array(Buffer.from(compiledClear.result, 'base64')),
        numLocalInts: 2,
        numLocalByteSlices: 2,
        numGlobalInts: 2,
        numGlobalByteSlices: 2,
    };

    const txn = makeApplicationCreateTxnFromObject(txnObject);

    // 5. Sign and Send using AlgoKit
    console.log("Sending and confirming transaction on TestNet...");
    const signedTxn = txn.signTxn(account.sk);
    const result = await algodClient.sendRawTransaction(signedTxn).do();

    console.log(`Transaction Sent: ${result.txId}`);

    // Wait for confirmation securely
    const confirmation = await algokit.waitForConfirmation(result.txId, 4, algodClient);
    
    console.log("\n🚀 CONTRACT SUCCESSFULLY DEPLOYED TO TESTNET 🚀");
    console.log(`App ID: ${confirmation['application-index']}`);
}

deployContracts().catch(console.error);
